File: 1-Part/6-chapter/combinations.js
function getCombos(chars, k, indent) {
    if (indent === undefined) {
        indent = 0
    }
    let debugMsg = ".".repeat(indent) + `In getCombos(${chars}, ${k})`
    console.log(debugMsg, ', start.')

    if (k === 0) {
        console.log(`${debugMsg} base case returns ['']`)
        return ['']
    } else if (chars === '') {
        console.log(`${debugMsg} base case returns []`)
        return []
    }

    let combinations = []
    let head = chars.slice(0, 1)
    let tail = chars.slice(1, chars.length)
    console.log(`${debugMsg} part 1, get combos with head ${head}`)
    let tailCombos = getCombos(tail, k - 1, indent + 1)
    console.log(`${".".repeat(indent)} Adding head ${head} to tail combos:`)
    for (let tailCombo of tailCombos) {
        console.log(`${".".repeat(indent)} New combination ${head} ${tailCombo}`)
        combinations.push(head + tailCombo)
    }

    console.log(`${debugMsg} part-2, get combos without head ${head}`)
    combinations = combinations.concat(getCombos(tail, k, indent + 1))

    console.log(`${debugMsg} results are ${combinations}`)
    return combinations
}

console.log('2-combinations of "ABC":')
console.log('Results:', getCombos('ABC', 2))

File: 1-Part/6-chapter/combinations.py
def getCombos(chars, k, indent=0):
    debugMsg = '.' * indent + "In getCombos('" + chars + "', " + str(k) + ")"
    print(debugMsg + ', start.')

    if k == 0:
        print(debugMsg + "base case returns []")
        return ['']
    elif chars == '':
        print(debugMsg + ' base case returns []')
        return []

    combinations = []

    head = chars[:1]
    tail = chars[1:]
    print(debugMsg + " part 1, get combos with head '" + head + "'")
    tailCombos = getCombos(tail, k - 1, indent + 1)
    print('.' * indent + "Adding head '" + head + "' to tail combos:")
    for tailCombo in tailCombos:
        print('.' * indent + 'New combination', head + tailCombo)
        combinations.append(head + tailCombo)

    combinations.extend(getCombos(tail, k, indent + 1))

    print(debugMsg + 'results are', combinations)
    return combinations

print('2-combinations of "ABC":')
print('Results:', getCombos('ABC', 2))
